crypto/src/kdf/mod.rs

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::num::NonZeroUsize;
+
+use crate::{
+    random::{CryptoRng, Rng},
+    util::eq::SliceEqualityCheckMethod,
+};
+use serialization::{Decode, Encode};
+
+pub mod argon2;
+
+#[derive(thiserror::Error, Debug, PartialEq, Eq, Clone)]
+pub enum KdfError {
+    #[error("Argon2 hashing error: {0}")]
+    Argon2HashingFailed(#[from] ::argon2::Error),
+    #[error("Invalid salt size")]
+    InvalidSaltSize,
+    #[error("Invalid hash size")]
+    InvalidHashSize,
+}
+
+#[derive(Clone, Debug)]
+pub enum KdfKind {
+    Argon2id {
+        m_cost_memory_size: u32,
+        t_cost_iterations: u32,
+        p_cost_parallelism: u32,
+        hash_length: NonZeroUsize,
+        salt_length: NonZeroUsize,
+    },
+}
+
+#[derive(Clone, Debug, Encode, Decode)]
+pub enum KdfResult {
+    #[codec(index = 0)]
+    Argon2id {
+        #[codec(compact)]
+        m_cost_memory_size: u32,
+        #[codec(compact)]
+        t_cost_iterations: u32,
+        #[codec(compact)]
+        p_cost_parallelism: u32,
+        salt: Vec<u8>,
+        hashed_password: Vec<u8>,
```

Comment:
Probably we would need to remove the `hashed_password` from here.

Reply:
hmmmm... you're right. I seem to have confused web authentication and wallet authentication. Let me think about it.